### PR TITLE
Issue 1024: Message Router should take `numPartitions` as parameter in choosing partition

### PR DESCRIPTION
**Pull request by sijie**
### Motivation

currently message router doesn't expose number partition. it is hard for customized message router to get such information. It would be good to expose this information in message router.

### Modifications

- introduce `TopicMetadata`
- use `TopicMetadata` in message router for choosing partitions

### Result

- deprecate the old `choosePartition` method
- add test cases for all the message routers

### CODE REVIEW

**Comment by sijie** [comment]
retest this please

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java [hunk 1]
```diff
@@ -40,21 +44,45 @@
 
 public class PartitionedProducerImpl extends ProducerBase {
 
+    private static final Random RANDOM = new Random(System.currentTimeMillis());
```
Comment (rdhabalia) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java]: if `PartitionedProducerImpl` is using it only once then should we need it as a class variable? if not then can we just keep its scope at line ln#80.?

Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java]: I was thinking about the case when you have a client producing to multiple topics, you will end up creating multiple random instances. I can switch to use ThreadLocalRandom if there is a concern using a static instance.

**Comment by sijie** [comment]
> I couldn't see any change regarding exposing partitions in MessageRouter.java file? I couldn't fully understand the change. can you provide a usecase for it?

@rdhabalia 

the partitions now is exposed at `TopicMetadata`, which is available at `#choosePartitions(msg, topicMetadata)`.

an example of customized message router would be : 

```
class SequenceIdMessageRouter implements MessageRouter {

     public int choosePartition(Message msg, TopicMetadata metadata) {
          return msg.getSequenceId() % metadata.numPartitions(); 
    }

}
```

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java (hunk 1 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java]: @rdhabalia I changed to use `ThreadLocalRandom`. let me know if it is okay or not.

**Comment by Licht-T** [comment]
@sijie Is there a plan to implement `TopicMetadata` on C++ client in this PR? If not, I'll do that.
I want to resolve https://github.com/apache/incubator-pulsar/pull/1029 after `TopicMetadata` is implemented on both clients.

**Comment by sijie** [comment]
@Licht-T I haven't started any work on implementing `TopicMetadata` on C++ client. Feel free to work on it.

**Comment by merlimat** [comment]
@Licht-T It would be nice to have the same info in the c++ router as well (and exposed in the Python binding too). The tricky part might be in doing so without breaking the compatibility with existing code.

**Comment by sijie** [comment]
retest this please

**Comment by Licht-T** [comment]
@sijie @merlimat Okay, I'll do that.

**Comment by sijie** [comment]
@rdhabalia can you review my last comment? If it is okay with you, can we merge this one? so I can be unblocked on using this feature :)

**Pull request merged**